py/profiling.c

```diff
@@ -0,0 +1,1046 @@
+    );
+
+    ip = prelude->locals;
+    while (*ip++ != 255);
+    prelude->bytecode = ip;
+}
+
+
+//
+// CODE object
+//
+
+STATIC void code_print(const mp_print_t *print, mp_obj_t o_in, mp_print_kind_t kind) {
+    (void)kind;
+    mp_obj_code_t *o = MP_OBJ_TO_PTR(o_in);
+    const mp_raw_code_t *rc = o->rc;
+    const mp_bytecode_prelude_t *prelude = &rc->prelude;
+    mp_printf(print,
+        "<code object %q at 0x%p, file \"%q\", line %d>",
+        prelude->qstr_block_name,
+        o,
+        prelude->qstr_source_file,
+        rc->line_of_definition
+    );
+}
+
+STATIC mp_obj_tuple_t* code_consts(const mp_raw_code_t *rc) {
+
+    const mp_bytecode_prelude_t *prelude = &rc->prelude;
+    int start = prelude->n_pos_args + prelude->n_kwonly_args + rc->n_obj;
+    int stop  = prelude->n_pos_args + prelude->n_kwonly_args + rc->n_obj + rc->n_raw_code;
+    mp_obj_tuple_t *consts = mp_obj_new_tuple(stop - start + 1, NULL);
+    size_t const_no = 0;
+    int i = 0;
+
+    // start = 0;
+    // stop  = prelude->n_pos_args + prelude->n_kwonly_args;
+    // for (i = start; i < stop; i++ ) {
+    //     consts->items[const_no++] = (mp_obj_t)rc->const_table[i];
+    // }
+
+    // start = prelude->n_pos_args + prelude->n_kwonly_args;
+    // stop  = prelude->n_pos_args + prelude->n_kwonly_args + rc->n_obj;
+    // for (i = start; i < stop; i++ ) {
+    //     consts->items[const_no++] = (mp_obj_t)rc->const_table[i];
+    // }
+
+    // start = prelude->n_pos_args + prelude->n_kwonly_args + rc->n_obj;
+    // stop  = prelude->n_pos_args + prelude->n_kwonly_args + rc->n_obj + rc->n_raw_code;
+    for (i = start; i < stop; i++ ) {
+        consts->items[const_no++] = mp_obj_new_code((const mp_raw_code_t*)rc->const_table[i]);
+    }
+
+    consts->items[const_no++] = mp_const_none;
+
+    return consts;
+}
+
+STATIC mp_obj_t raw_code_lnotab(const mp_raw_code_t *rc) {
+    // const mp_bytecode_prelude_t *prelude = &rc->prelude;
+    uint start = 0;
+    uint stop = rc->fun_data_len - start;
+
+    uint last_lineno = prof_bytecode_lineno(rc, start);
+    uint lasti = 0;
+
+    const uint buffer_chunk_size = (stop-start) >> 2; // heuristic magic
+    uint buffer_size = buffer_chunk_size;
+    byte *buffer = m_new(byte, buffer_size);
+    uint buffer_index = 0;
+
+    int i;
+    for (i = start; i < stop; i ++) {
+        uint lineno = prof_bytecode_lineno(rc, i);
+        size_t line_diff = lineno - last_lineno;
+        if (line_diff > 0) {
+            uint instr_diff = (i - start) - lasti;
+            
+            assert(instr_diff < 256);
+            assert(line_diff < 256);
+
+            // mp_printf(&mp_plat_print, "lnotab %d, %d\n", instr_diff, line_diff);
+            if (buffer_index + 2 > buffer_size) {
+                // mp_printf(&mp_plat_print, "has to realloc lnotab\n");
+                #if MICROPY_MALLOC_USES_ALLOCATED_SIZE
+                buffer = m_realloc(buffer, buffer_size, buffer_size + buffer_chunk_size);
```

Comment:
You should be able to use `m_renew(byte, buffer, buffer_size, buffer_size + buffer_chunk_size)`, then no need for a #if.